src/ray/common/status_or.h

```diff
@@ -0,0 +1,261 @@
+  auto and_then(F &&f) const &;
+  template <typename F>
+  auto and_then(F &&f) &&;
+
+  // Returns a reference to the current value.
+  //
+  // REQUIRES: `this->ok() == true`, otherwise the behavior is undefined.
+  //
+  // Use `this->ok()` to verify that there is a current value within the
+  // `ray::StatusOr<T>`. Alternatively, see the `value()` member function for a
+  // similar API that guarantees crashing or throwing an exception if there is
+  // no current value.
+  T &operator*() & { return get(); }
+  const T &operator*() const & { return get(); }
+  T &&operator*() && { return std::move(get()); }
+
+  // Returns a pointer to the current value.
+  //
+  // REQUIRES: `this->ok() == true`, otherwise the behavior is undefined.
+  //
+  // Use `this->ok()` to verify that there is a current value.
+  T *operator->() & { return &data_; }
+  T *operator->() const & { return &data_; }
+
+  // Returns a reference to the held value if `this->ok()`. Otherwise, throws
+  // `std::runtime_error`.
+  //
+  // If you have already checked the status using `this->ok()`, you probably
+  // want to use `operator*()` or `operator->()` to access the value instead of
+  // `value`.
+  //
+  // Note: for value types that are cheap to copy, prefer simple code:
+  //
+  //   T value = statusor.value();
+  //
+  // Otherwise, if the value type is expensive to copy, but can be left
+  // in the StatusOr, simply assign to a reference:
+  //
+  //   T& value = statusor.value();  // or `const T&`
+  //
+  // Otherwise, if the value type supports an efficient move, it can be
+  // used as follows:
+  //
+  //   T value = std::move(statusor).value();
+  //
+  // The `std::move` on statusor instead of on the whole expression enables
+  // warnings about possible uses of the statusor object after the move.
+  T &value() &;
+  const T &value() const &;
+  T &&value() &&;
+
+  void swap(StatusOr &rhs);
+
+  // Copy current value out if OK status, otherwise construct default value.
+  T value_or_default() const & {
+    static_assert(std::is_copy_constructible_v<T>, "T must by copy constructable");
+    static_assert(std::is_default_constructible_v<T>, "T must by default constructable");
+    if (ok()) return get();
+    return T{};
+  }
+  T value_or_default() && {
+    static_assert(std::is_copy_constructible_v<T>, "T must by copy constructable");
+    static_assert(std::is_default_constructible_v<T>, "T must by default constructable");
+    if (ok()) return std::move(get());
+    return T{};
+  }
+
+  static_assert(std::is_default_constructible_v<T>,
+                "StatusOr<T>::value_or_default: T must by default constructable");
+
+ private:
+  T &get() { return data_; }
+  const T &get() const { return data_; }
+
+  void CheckHasValue() const {
+    if (!has_value()) {
+      throw std::runtime_error(status_.message());
+    }
+  }
+
+  Status status_;
+  T data_;
+};
+
+template <typename T>
+template <typename F>
+auto StatusOr<T>::and_then(F &&f) & {
```

Comment:
Q: do we want this? In what case do we want to have a & this, other than a const& or a && ?

Reply:
`std::forward<F>(f)(*this)` functor takes a mutable reference of `statusor`.

Reply:
I mean, maybe only &&this is needed?

Reply:
No, rvalue overload indicates ownership transfer; while non-const lvalue overload only for mutability.

Reply:
i think deducing this would figure this out, time to go straight to c++23 🚀